cmd/permissions-checker/permissionschecks.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os/user"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"syscall"
+)
+
+type currentPermissions struct {
+	permissions           string
+	defaultPermissions    string
+	owner                 string
+	group                 string
+	hasCorrectPermissions bool
+	hasCorrectOwner       bool
+	hasCorrectGroup       bool
+}
+
+// FindRecursive iterates in a directory showing permissions in a recursive way
+/* path [string] 											- where to look for files and directories
+ * defaultPerm [defaultPermissions] 	- data regarding default permissions
+ * 	defaultPerm.file [string]					- default permissions for files
+ * 	defaultPerm.dir [string] 					- default permissions for directories
+ * 	defaultPerm.owner [string] 				- default owner
+ * 	defaultPerm.group [string] 				- default group
+ * search [searchSettings] 						- data regarding search options
+ * 	search.hidden [bool] 							- includes hidden files and directories in the search
+ * 	search.exclude [string] 					- files and/or directories to be excluded (regexp)
+ * 	search.baseDirectory [string]			- base directory
+ * level [int]                        - used to print the output in a hierarchical way
+ */
+func FindRecursive(path string, defaultPerm defaultPermissions, search searchSettings, verbose bool, level int) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// In verbose mode, we are printing the output in a hierarchical way
+	if verbose {
+		level++
+	}
+
+	for _, f := range files {
+		name := f.Name()
+
+		relativePath := strings.TrimPrefix(strings.Join([]string{path, name}, "/"), search.baseDirectory)
+		fullPath := strings.Join([]string{path, name}, "/")
+
+		if m, _ := regexp.MatchString(search.exclude, relativePath); m {
+			if verbose {
+				fmt.Printf("\x1b[33m%sExcluding %s\n\x1b[0m", strings.Repeat(" ", level), name)
+			}
+		} else {
+			if search.hidden || !strings.HasPrefix(name, ".") {
+				mode := f.Mode()
+				currentOwner, _ := user.LookupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid))
+				currentGroup, _ := user.LookupGroupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Gid))
+
+				currentPerm := currentPermissions{
+					permissions:     mode.Perm().String(),
```

Comment:
why are we losing the strong typed https://golang.org/pkg/os/#FileMode type and converting it to string so eagerly here.

Keep the `os.FileMode` type instead of `rwxrwxrwx` strings all over the place would make everything a bit more self-documenting and safe.